File: src/services/docker/composeRunner.ts
import { spawn as realSpawn } from "node:child_process";

export interface Spawner {
  spawn(
    cmd: string,
    args: string[],
    opts: { cwd: string; signal?: AbortSignal },
  ): AsyncGenerator<string, number>;
}

// Exported so the abort-ordering behavior can be unit-tested directly.
export const defaultSpawner: Spawner = {
  spawn: async function* spawn(cmd, args, opts) {
    const child = realSpawn(cmd, args, { cwd: opts.cwd, stdio: ["ignore", "pipe", "pipe"] });
    const out = child.stdout;
    const err = child.stderr;
    out.setEncoding("utf-8");
    err.setEncoding("utf-8");
    let lineBuffer = "";
    let resolveChunk: ((v: string | null) => void) | null = null;
    const chunkQueue: Array<string | null> = [];
    const pushChunk = (chunk: string | null) => {
      if (resolveChunk) {
        resolveChunk(chunk);
        resolveChunk = null;
      } else {
        chunkQueue.push(chunk);
      }
    };
    // Register stream + close listeners FIRST so an already-aborted signal
    // (handled below) cannot kill the child before `close` is observable.
    out.on("data", (chunk: string) => pushChunk(chunk));
    err.on("data", (chunk: string) => pushChunk(chunk));
    const exitPromise = new Promise<number>((res) =>
      child.on("close", (code: number | null) => {
        pushChunk(null);
        res(code ?? 0);
      }),
    );
    const onAbort = () => {
      child.kill("SIGTERM");
      pushChunk(null);
    };
    if (opts.signal) {
      if (opts.signal.aborted) onAbort();
      else opts.signal.addEventListener("abort", onAbort, { once: true });
    }
    try {
      while (true) {
        let chunk: string | null;
        if (chunkQueue.length > 0) {
          chunk = chunkQueue.shift() ?? null;
        } else {
          chunk = await new Promise<string | null>((res) => {
            resolveChunk = res;
          });
        }
        if (chunk === null) {
          if (lineBuffer.length > 0) yield lineBuffer;
          break;
        }
        lineBuffer += chunk;
        const lines = lineBuffer.split("\n");
        lineBuffer = lines.pop() ?? "";
        for (const line of lines) {
          yield `${line}\n`;
        }
      }
    } finally {
      opts.signal?.removeEventListener("abort", onAbort);
      // If the consumer stopped early (e.g. `break` out of the for-await without
      // aborting), the generator returns here and `await exitPromise` below is
      // skipped — so kill any still-running child to avoid leaking a `logs -f`
      // subprocess. No-op on the normal-close and abort paths (child already gone).
      if (child.exitCode === null && child.signalCode === null) child.kill("SIGTERM");
    }
    return await exitPromise;
  },
};

export interface UpOpts {
  detach?: boolean;
  scale?: Record<string, number>;
}
export interface DownOpts {
  volumes?: boolean;
}
export interface PsOpts {
  json?: boolean;
}
export interface LogsOpts {
  service?: string;
  tailLines?: number;
  follow?: boolean;
  since?: string;
  signal?: AbortSignal;
}
export interface ComposePsRow {
  Name: string;
  Service: string;
  State: string;
  Health?: string;
}

export class BoundComposeRunner {
  constructor(
    public readonly stackName: string,
    public readonly yamlPath: string,
    public readonly cwd: string,
    private spawner: Spawner,
  ) {}

  private baseArgs(): string[] {
    return ["compose", "-p", this.stackName, "--project-directory", this.cwd, "-f", this.yamlPath];
  }

  async *up(opts: UpOpts = {}): AsyncGenerator<string, number> {
    const args = this.baseArgs();
    args.push("up");
    if (opts.detach) args.push("-d");
    for (const [svc, n] of Object.entries(opts.scale ?? {})) args.push("--scale", `${svc}=${n}`);
    return yield* this.spawner.spawn("docker", args, { cwd: this.cwd });
  }

  async *down(opts: DownOpts = {}): AsyncGenerator<string, number> {
    const args = this.baseArgs();
    args.push("down");
    if (opts.volumes) args.push("-v");
    return yield* this.spawner.spawn("docker", args, { cwd: this.cwd });
  }

  async ps(opts: PsOpts = {}): Promise<ComposePsRow[]> {
    const args = this.baseArgs();
    args.push("ps");
    if (opts.json) args.push("--format", "json");
    const rows: ComposePsRow[] = [];
    let acc = "";
    const gen = this.spawner.spawn("docker", args, { cwd: this.cwd });
    while (true) {
      const r = await gen.next();
      if (r.done) break;
      acc += r.value;
    }
    for (const line of acc.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      try {
        rows.push(JSON.parse(trimmed) as ComposePsRow);
      } catch {
        /* ignore non-JSON */
      }
    }
    return rows;
  }

  async *logs(opts: LogsOpts = {}): AsyncGenerator<string, number> {
    const args = this.baseArgs();
    args.push("logs");
    if (opts.follow) args.push("-f");
    if (opts.tailLines !== undefined) args.push("--tail", String(opts.tailLines));
    if (opts.since !== undefined) args.push("--since", opts.since);
    if (opts.service) args.push(opts.service);
    return yield* this.spawner.spawn("docker", args, {
      cwd: this.cwd,
      ...(opts.signal ? { signal: opts.signal } : {}),
    });
  }
}

export class ComposeRunner {
  constructor(
    private cwd: string,
    private spawner: Spawner = defaultSpawner,
  ) {}
  forStack(stackName: string, yamlPath: string): BoundComposeRunner {
    return new BoundComposeRunner(stackName, yamlPath, this.cwd, this.spawner);
  }
}

File: python/src/docker_agent/services/docker/compose_runner.py
"""Subprocess runner for ``docker compose`` commands.

Parity: ``src/services/docker/composeRunner.ts:1-178``.
"""

from __future__ import annotations

import asyncio
import json as json_module
from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any, Protocol


class Spawner(Protocol):
    def spawn(
        self, cmd: str, args: list[str], opts: dict[str, Any]
    ) -> AsyncIterator[str]: ...


async def _default_spawner_impl(
    cmd: str, args: list[str], opts: dict[str, Any]
) -> AsyncIterator[str]:
    """Default subprocess spawner using ``asyncio.create_subprocess_exec``."""
    cwd = opts["cwd"]
    signal = opts.get("signal")

    proc = await asyncio.create_subprocess_exec(
        cmd,
        *args,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    abort_task: asyncio.Task[Any] | None = None

    def on_abort() -> None:
        if proc.returncode is None:
            proc.terminate()

    if signal is not None:
        abort_task = asyncio.create_task(signal.wait())
        abort_task.add_done_callback(lambda _: on_abort())

    try:
        stdout = proc.stdout
        stderr = proc.stderr
        assert stdout is not None and stderr is not None

        async def read_all(reader: asyncio.StreamReader) -> list[str]:
            chunks: list[str] = []
            while True:
                line = await reader.readline()
                if not line:
                    break
                chunks.append(line.decode("utf-8", errors="replace"))
            return chunks

        stdout_chunks, stderr_chunks = await asyncio.gather(
            read_all(stdout), read_all(stderr)
        )
        line_buffer = ""
        for text in stdout_chunks + stderr_chunks:
            line_buffer += text
            while "\n" in line_buffer:
                out, line_buffer = line_buffer.split("\n", 1)
                yield out.rstrip("\n")
        if line_buffer:
            yield line_buffer.rstrip("\n")
    finally:
        if abort_task is not None and not abort_task.done():
            abort_task.cancel()
        if proc.returncode is None:
            proc.terminate()
        await proc.wait()


class DefaultSpawner:
    def spawn(
        self, cmd: str, args: list[str], opts: dict[str, Any]
    ) -> AsyncIterator[str]:
        return _default_spawner_impl(cmd, args, opts)


default_spawner = DefaultSpawner()


@dataclass(frozen=True)
class ComposePsRow:
    name: str
    service: str
    state: str
    health: str | None = None


class BoundComposeRunner:
    """A compose runner bound to one stack + yaml path."""

    def __init__(
        self,
        stack_name: str,
        yaml_path: str,
        cwd: str,
        spawner: Spawner,
    ) -> None:
        self.stack_name = stack_name
        self.yaml_path = yaml_path
        self.cwd = cwd
        self.spawner = spawner

    def base_args(self) -> list[str]:
        return [
            "compose",
            "-p",
            self.stack_name,
            "--project-directory",
            self.cwd,
            "-f",
            self.yaml_path,
        ]

    async def up(
        self, *, detach: bool = False, scale: dict[str, int] | None = None
    ) -> AsyncIterator[str]:
        args = self.base_args() + ["up"]
        if detach:
            args.append("-d")
        for svc, n in sorted((scale or {}).items()):
            args.extend(["--scale", f"{svc}={n}"])
        async for line in self.spawner.spawn("docker", args, {"cwd": self.cwd}):
            yield line

    async def down(self, *, volumes: bool = False) -> AsyncIterator[str]:
        args = self.base_args() + ["down"]
        if volumes:
            args.append("-v")
        async for line in self.spawner.spawn("docker", args, {"cwd": self.cwd}):
            yield line

    async def ps(self, *, json: bool = False) -> list[ComposePsRow]:
        args = self.base_args() + ["ps"]
        if json:
            args.extend(["--format", "json"])
        rows: list[ComposePsRow] = []
        async for line in self.spawner.spawn("docker", args, {"cwd": self.cwd}):
            line = line.strip()
            if not line:
                continue
            try:
                data = json_module.loads(line)
                rows.append(
                    ComposePsRow(
                        name=data["Name"],
                        service=data["Service"],
                        state=data["State"],
                        health=data.get("Health"),
                    )
                )
            except Exception:  # noqa: BLE001
                pass
        return rows

    async def logs(
        self,
        *,
        service: str | None = None,
        tail_lines: int | None = None,
        follow: bool = False,
        since: str | None = None,
        signal: Any | None = None,
    ) -> AsyncIterator[str]:
        args = self.base_args() + ["logs"]
        if follow:
            args.append("-f")
        if tail_lines is not None:
            args.extend(["--tail", str(tail_lines)])
        if since is not None:
            args.extend(["--since", since])
        if service is not None:
            args.append(service)
        async for line in self.spawner.spawn(
            "docker", args, {"cwd": self.cwd, "signal": signal}
        ):
            yield line


class ComposeRunner:
    """Factory for ``BoundComposeRunner``."""

    def __init__(self, cwd: str, spawner: Spawner | None = None) -> None:
        self.cwd = cwd
        self.spawner = spawner if spawner is not None else default_spawner

    def for_stack(self, stack_name: str, yaml_path: str) -> BoundComposeRunner:
        return BoundComposeRunner(stack_name, yaml_path, self.cwd, self.spawner)


__all__ = [
    "BoundComposeRunner",
    "ComposePsRow",
    "ComposeRunner",
    "DefaultSpawner",
    "Spawner",
    "default_spawner",
]